src/views/projects/SourceBrowser/FileDiff.svelte

```diff
@@ -58,6 +72,101 @@
       }
     }
   }
+
+  function thisLineSelected(
+    uriHunkLine: string | undefined,
+    hunkIdx: number,
+    lineIdx: number,
+  ): boolean {
+    if (!isSelected || !uriHunkLine) {
+      return false;
+    }
+
+    // range or single selection
+    const match = uriHunkLine.match(/H(\d*)L(\d*)L?(\d*)/);
+    if (match) {
+      if (match[3]) {
+        return (
+          match[1] === hunkIdx.toString() &&
+          lineIdx >= parseInt(match[2]) &&
+          lineIdx <= parseInt(match[3])
+        );
+      } else {
+        return uriHunkLine === "H" + hunkIdx + "L" + lineIdx;
+      }
+    }
+
+    return false;
+  }
+
+  function shouldScroll(
+    uriHunkLine: string | undefined,
+    hunkIdx: number,
+    lineIdx: number,
+  ): boolean {
+    if (!isSelected || !uriHunkLine) {
+      return false;
+    }
+
+    // range or single selection
+    const match = uriHunkLine.match(/H(\d*)L(\d*)L?(\d*)/);
+    if (match) {
+      if (match[3]) {
+        return (
+          match[1] === hunkIdx.toString() && lineIdx.toString() === match[2]
+        );
+      } else {
+        return uriHunkLine === "H" + hunkIdx + "L" + lineIdx;
+      }
+    }
+
+    return false;
+  }
+
+  function calcNewHash(
+    path: string,
+    hunkIdx: number,
+    lineIdx: number,
+    shift: boolean,
+  ): string {
+    // single line selection
+    if (!shift) {
+      return path + ":H" + hunkIdx + "L" + lineIdx;
+    }
+
+    if (!selectedHunkLine) {
+      return "";
+    }
+
+    // range selection
+    const match = selectedHunkLine.match(/H(\d*)L(\d*)L?(\d*)/);
+    if (match) {
+      // make sure we're on the same hunk
+      if (match[1] === hunkIdx.toString()) {
+        if (lineIdx >= parseInt(match[2])) {
+          return path + ":H" + hunkIdx + "L" + match[2] + "L" + lineIdx;
+        } else {
+          return path + ":H" + hunkIdx + "L" + lineIdx + "L" + match[2];
+        }
+      }
+    }
+
+    return "";
+  }
+
+  function updateHash(newHash: string) {
+    if (newHash !== "") {
+      window.location.hash = newHash;
+    } else {
+      // eslint-disable-next-line no-restricted-globals
+      history.pushState(
```

Comment:
`npm run check` forced me to add `// eslint...` otherwise:

```
error  Unexpected use of 'history'  no-restricted-globals
```

--

had to `history` here because, if we just set the hash to `""`, it'll still trigger a scroll to `link#`, which is the top of the page, both annoying and undesired. with `history` we can remove the hash, with forcing a reload of the page.

Reply:
You can use `window.history`. I’m curious why you need this code

> had to use `history` here because, if we just set the hash to `""`, it'll still trigger a scroll to `<link>#`,

Can you explain which actions the user needs to take to trigger this  code path? I wasn’t able to do it because `newHash` is always set when I try this.

Reply:
If you start a selection and then cross hunk boundary, on click it'd return `""` and hit the code path.

Reply:
Ah, yes.

I’d suggest the following adjustment to retain the state and also pass an empty string as the second argument. This argument is ignored by browsers and passing `document.title` is a legacy thing.

```typescript
      window.history.replaceState(
        window.history.state,
        "",
        window.location.pathname + window.location.search,
      );
```